### PR TITLE
Refactor Error and Result

### PR DESCRIPTION
**Pull request by lthiery**

### CODE REVIEW

File: src/client_runtime/error.rs [hunk 1]
```diff
@@ -0,0 +1,29 @@
+use thiserror::Error;
+use tokio::sync::mpsc;
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("semtech udp error")]
+    SemtechUdpError(crate::packet::Error),
```
Comment (madninja) [inline src/client_runtime/error.rs:7]: I've come to remove the `Error` suffix in the enum itself because they're already part of that top level `Error`

File: src/client_runtime/error.rs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    #[error("tokio::mpsc send error")]
+    SendError(mpsc::error::SendError<super::TxMessage>),
+    #[error("std::io::Error")]
+    IoError(std::io::Error),
+}
+
+impl From<crate::packet::Error> for Error {
```
Comment (madninja) [inline src/client_runtime/error.rs:14]: Why can't you use the `#[from]` annotation in the enum above for all of these From implementations?

File: src/client_runtime/mod.rs [hunk 3]
```diff
@@ -3,7 +3,7 @@
    run sending and receiving concurrently as tasks,
    receive downlink packets and send uplink packets easily
 */
-use super::{parser::Parser, pull_data, Down, MacAddress, Packet, SerializablePacket, Up};
+use crate::{parser::Parser, pull_data, Down, MacAddress, Packet, SerializablePacket, Up};
```
Comment (madninja) [inline src/client_runtime/mod.rs:6]: I've resorted to using `crate::*` in my own crates now to make for less noisy source files. Personal preference

File: src/client_runtime/mod.rs [hunk 4]
```diff
@@ -72,7 +51,7 @@ impl UdpRuntime {
         self.rx.sender.subscribe()
     }
 
-    pub async fn run(self) -> Result<(), Error> {
+    pub async fn run(self) -> Result<()> {
```
Comment (madninja) [inline src/client_runtime/mod.rs:54]: Remove the `<()>` since your `Result` type already has this as the default

File: src/client_runtime/mod.rs [hunk 5]
```diff
@@ -12,6 +12,10 @@ use tokio::sync::{
     mpsc::{self, Receiver, Sender},
 };
 
+mod error;
+pub use error::Error;
+pub type Result<T = ()> = std::result::Result<T, Error>;
```
Comment (madninja) [inline src/client_runtime/mod.rs:17]: Why is this here when you have one already defined in the top level lib?

File: src/packet/error.rs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    InvalidProtocolVersion,
+    #[error("invalid GWMP frame identifier")]
+    InvalidIdentifier,
+    #[error("utf8 error")]
+    Utf8(std::str::Utf8Error),
+    #[error("unable to parse GWMP JSON")]
+    Json,
+}
+
+impl From<ParseError> for Error {
```
Comment (madninja) [inline src/packet/error.rs:25]: You should be able to use the `#[from]` annotation and remove all of these from implementations

File: src/server_runtime/error.rs [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    InternalQueueClosedOrFull,
+    #[error("Semtech UDP error")]
+    SemtechUdp(crate::packet::Error),
+    #[error("error receiving ACK")]
+    AckRecvError,
+    #[error("error sending ACK")]
+    ErrorSendingAck,
+}
+
+impl From<tokio::time::error::Elapsed> for Error {
```
Comment (madninja) [inline src/server_runtime/error.rs:31]: Use the `#[from]` force, Luke

File: src/server_runtime/mod.rs [hunk 8]
```diff
@@ -12,6 +12,10 @@ use tokio::{
 
 pub use crate::push_data::RxPk;
 
+mod error;
+pub use error::Error;
+pub type Result<T = ()> = std::result::Result<T, Error>;
```
Comment (madninja) [inline src/server_runtime/mod.rs:17]: I don't think you need to define this again, do you? I see you have runtime errors as their own separate types, but why not embody that all in the top level Error with some sub enum types that are pulled in as needed?

File: src/client_runtime/mod.rs (hunk 5 above)
Comment (lthiery) [reply src/client_runtime/mod.rs:17]: in this case the top-level Error is different. It's client_runtime::Error which contains the packet::Error (which is the base of the crate) as a sub-Error. But I think I _will_ move the Result definition to the packet module rather than `lib.rs`

File: src/client_runtime/mod.rs (hunk 3 above)
Comment (lthiery) [reply src/client_runtime/mod.rs:6]: I'm avoiding it in this case to avoid the packet::Error/Result types from taking those names from me locally
